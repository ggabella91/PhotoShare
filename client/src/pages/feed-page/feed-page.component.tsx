import React, { useState, useEffect, useRef, useCallback } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { createStructuredSelector } from 'reselect';

import { AppState } from '../../redux/root-reducer';

import {
  User,
  OtherUserRequest,
  OtherUserType,
} from '../../redux/user/user.types';
import {
  selectCurrentUser,
  selectFollowingInfo,
} from '../../redux/user/user.selectors';
import {
  getOtherUserStart,
  clearFollowersAndFollowing,
} from '../../redux/user/user.actions';

import {
  Post,
  DataRequestType,
  PostDataReq,
  PostFileReq,
  PostFile,
  PostError,
  UserType,
  PostMetaData,
} from '../../redux/post/post.types';
import {
  selectPostDataFeedArray,
  selectPostFiles,
  selectFollowPhotoFileArray,
  selectPostError,
  selectPostConfirm,
  selectGetPostDataConfirm,
  selectGetPostDataError,
  selectGetPostFileConfirm,
  selectGetPostFileError,
  selectIsLoadingPostData,
  selectPostMetaDataForUser,
} from '../../redux/post/post.selectors';
import {
  getPostDataStart,
  getPostFileStart,
  archivePostStart,
  clearFollowPhotoFileArray,
  clearPostState,
} from '../../redux/post/post.actions';

import {
  Follower,
  FollowError,
  WhoseUsersFollowing,
  UsersFollowingRequest,
} from '../../redux/follower/follower.types';
import {
  selectCurrentUserUsersFollowing,
  selectGetUsersFollowingConfirm,
} from '../../redux/follower/follower.selectors';
import {
  getUsersFollowingStart,
  clearFollowState,
} from '../../redux/follower/follower.actions';

import FeedPostContainer from '../../components/feed-post-container/feed-post-container.component';

import { prepareUserInfoAndFileArray } from './feed-page.utils';
import './feed-page.styles.scss';

export interface PostDataArrayMap {
  postData: Post[];
  queryLength?: number;
  userId: string;
}

export interface UserInfoAndPostFile {
  profilePhotoFileString: string;
  username: string;
  userId: string;
  location: string;
  postId: string;
  postFileString: string;
  caption?: string;
  dateString: string;
  dateInt: number;
}

interface FeedPageProps {
  currentUser: User | null;
  postDataFeedArray: Post[][];
  postFiles: PostFile[];
  postConfirm: string | null;
  postError: PostError | null;
  getPostDataConfirm: string | null;
  getPostDataError: PostError | null;
  getPostFileConfirm: string | null;
  getPostFileError: PostError | null;
  currentUserUsersFollowing: Follower[] | null;
  followingInfo: User[] | null;
  followPhotoFileArray: PostFile[] | null;
  getUsersFollowingConfirm: string | null;
  isLoadingPostData: boolean;
  postMetaDataForUser: PostMetaData | null;
  getPostDataStart: typeof getPostDataStart;
  getPostFileStart: typeof getPostFileStart;
  clearPostState: typeof clearPostState;
  getUsersFollowingStart: typeof getUsersFollowingStart;
  getOtherUserStart: typeof getOtherUserStart;
  clearFollowersAndFollowing: typeof clearFollowersAndFollowing;
  clearFollowState: typeof clearFollowState;
}

export const FeedPage: React.FC<FeedPageProps> = ({
  currentUser,
  postDataFeedArray,
  postFiles,
  currentUserUsersFollowing,
  followingInfo,
  followPhotoFileArray,
  isLoadingPostData,
  postMetaDataForUser,
  getPostDataConfirm,
  getPostDataStart,
  getPostFileStart,
  clearPostState,
  getUsersFollowingStart,
  getOtherUserStart,
  clearFollowersAndFollowing,
  clearFollowState,
}) => {
  const [user, setUser] = useState({
    id: '',
    name: '',
    username: '',
    bio: '',
  });

  const [usersFollowingArray, setUsersFollowingArray] =
    useState<Follower[] | null>(null);

  const [followingInfoArray, setFollowingInfoArray] =
    useState<User[] | null>(null);

  const [dataFeedMapArray, setDataFeedMapArray] =
    useState<PostDataArrayMap[] | null>(null);

  const [followingProfilePhotoArray, setFollowingProfilePhotoArray] =
    useState<PostFile[] | null>(null);

  const [postFileFeedArray, setPostFileFeedArray] =
    useState<PostFile[] | null>(null);

  const [userInfoAndPostFileArray, setUserInfoAndPostFileArray] =
    useState<UserInfoAndPostFile[] | null>(null);

  const [pageToFetch, setPageToFetch] = useState(1);

  let postsBucket: string, profileBucket: string;

  if (process.env.NODE_ENV === 'production') {
    postsBucket = 'photo-share-app';
    profileBucket = 'photo-share-app-profile-photos';
  } else {
    postsBucket = 'photo-share-app-dev';
    profileBucket = 'photo-share-app-profile-photos-dev';
  }

  useEffect(() => {
    if (currentUser && !user.name) {
      clearPostState();
      clearFollowState();
      clearFollowersAndFollowing();

      setUser({
        id: currentUser.id,
        name: currentUser.name,
        username: currentUser.username,
        bio: currentUser.bio || '',
      });
      getUsersFollowingStart({
        userId: currentUser.id,
        whoseUsersFollowing: WhoseUsersFollowing.CURRENT_USER,
      });
    }
  }, [currentUser]);

  useEffect(() => {
    if (currentUserUsersFollowing) {
      setUsersFollowingArray(currentUserUsersFollowing);
    } else {
      setUsersFollowingArray(null);
    }
  }, [currentUserUsersFollowing]);

  useEffect(() => {
    if (usersFollowingArray) {
      for (let user of usersFollowingArray) {
        if (currentUser) {
          getOtherUserStart({
            usernameOrId: user.userId,
            type: OtherUserType.FOLLOWING,
          });

          getPostDataStart({
            userId: user.userId,
            dataReqType: DataRequestType.feed,
            pageToShow: pageToFetch,
            limit: 2,
          });
        }
      }
    }
  }, [usersFollowingArray]);

  useEffect(() => {
    if (postMetaDataForUser && dataFeedMapArray) {
      for (let el of dataFeedMapArray) {
        if (postMetaDataForUser.userId === el.userId) {
          el.queryLength = postMetaDataForUser.queryLength;
        }
      }

      // This assumes that properties in the dataFeedMapArray's
      // objects were successfully changed
      setDataFeedMapArray(dataFeedMapArray);
    }
  }, [postMetaDataForUser]);

  useEffect(() => {
    if (pageToFetch > 1 && dataFeedMapArray) {
      for (let el of dataFeedMapArray) {
        if (
          el.queryLength &&
          currentUser &&
          pageToFetch <= el.queryLength / 2
        ) {
          getPostDataStart({
            userId: el.userId,
            dataReqType: DataRequestType.feed,
            pageToShow: pageToFetch,
            limit: 2,
          });
        }
      }
    }
  }, [pageToFetch, dataFeedMapArray]);

  useEffect(() => {
    if (postDataFeedArray.length) {
      if (dataFeedMapArray) {
        for (let el of postDataFeedArray) {
          for (let mapEl of dataFeedMapArray) {
            if (el[0].userId === mapEl.userId) {
              mapEl.postData = el;
            }
          }
        }

        // This assumes that properties in the dataFeedMapArray's
        // objects were successfully changed
        setDataFeedMapArray(dataFeedMapArray);
      } else {
        let dataMapArray: PostDataArrayMap[] = [];

        for (let el of postDataFeedArray) {
          dataMapArray.push({ postData: el, userId: el[0].userId });
        }

        setDataFeedMapArray(dataMapArray);
      }
    }
  }, [postDataFeedArray]);

  useEffect(() => {
    if (followingInfo) {
      setFollowingInfoArray(followingInfo);
    }
  }, [followingInfo]);

  useEffect(() => {
    if (currentUser && followingInfoArray) {
      for (let el of followingInfoArray) {
        if (el.photo) {
          getPostFileStart({
            s3Key: el.photo,
            bucket: profileBucket,
            user: UserType.followArray,
          });
        }
      }
    }
  }, [followingInfoArray]);

  useEffect(() => {
    if (currentUser && dataFeedMapArray) {
      for (let innerObj of dataFeedMapArray) {
        for (let el of innerObj.postData) {
          getPostFileStart({
            s3Key: el.s3Key,
            bucket: postsBucket,
            user: UserType.other,
          });
        }
      }
    }
  }, [dataFeedMapArray, getPostDataConfirm]);

  useEffect(() => {
    if (followPhotoFileArray) {
      setFollowingProfilePhotoArray(followPhotoFileArray);
    }
  }, [followPhotoFileArray]);

  useEffect(() => {
    setPostFileFeedArray(postFiles);
  }, [postFiles]);

  useEffect(() => {
    if (
      followingInfoArray &&
      dataFeedMapArray &&
      followingProfilePhotoArray &&
      postFileFeedArray
    ) {
      let postDataArray: Post[][] = [];

      for (let el of dataFeedMapArray) {
        postDataArray.push(el.postData);
      }

      const userInfoAndPostObjArray = prepareUserInfoAndFileArray(
        followingInfoArray,
        postDataArray,
        followingProfilePhotoArray,
        postFileFeedArray
      );

      const sortedUserInfoAndPostArray = userInfoAndPostObjArray.sort(
        (a, b) => b.dateInt - a.dateInt
      );

      setUserInfoAndPostFileArray(sortedUserInfoAndPostArray);
    }
  }, [
    followingInfoArray,
    dataFeedMapArray,
    followingProfilePhotoArray,
    postFileFeedArray,
    getPostDataConfirm,
  ]);

  const observer = useRef<IntersectionObserver>();

  const lastPostContainerElementRef = useCallback(
    (node: HTMLDivElement | null) => {
      if (isLoadingPostData) return;

      if (observer.current) {
        observer.current.disconnect();
      }

      observer.current = new IntersectionObserver((entries) => {
        if (entries[0].isIntersecting) {
          setPageToFetch(pageToFetch + 1);
        }
      });

      if (node) {
        observer.current.observe(node);
      }
    },
    [isLoadingPostData]
  );

  return (
    <div className='feed-page'>
      {userInfoAndPostFileArray && userInfoAndPostFileArray.length ? (
        userInfoAndPostFileArray.map((el, idx) => {
          if (idx === userInfoAndPostFileArray.length - 1) {
            return (
              <FeedPostContainer
                userInfo={{
                  profilePhotoFileString: el.profilePhotoFileString,
                  username: el.username,
                  userId: el.userId,
                  postId: el.postId,
                  location: el.location,
                  name: '',
                  comment: '',
                }}
                fileString={el.postFileString}
                caption={el.caption}
                date={el.dateString}
                key={Math.random()}
                custRef={lastPostContainerElementRef}
              />
            );
          } else {
            return (
              <FeedPostContainer
                userInfo={{
                  profilePhotoFileString: el.profilePhotoFileString,
                  username: el.username,
                  userId: el.userId,
                  postId: el.postId,
                  location: el.location,
                  name: '',
                  comment: '',
                }}
                fileString={el.postFileString}
                caption={el.caption}
                date={el.dateString}
                key={Math.random()}
              />
            );
          }
        })
      ) : (
        <div className='no-franz'>
          Follow users to see their recent posts here
        </div>
      )}
    </div>
  );
};

interface LinkStateProps {
  currentUser: User | null;
  postDataFeedArray: Post[][];
  postFiles: PostFile[];
  postConfirm: string | null;
  postError: PostError | null;
  getPostDataConfirm: string | null;
  getPostDataError: PostError | null;
  getPostFileConfirm: string | null;
  getPostFileError: PostError | null;
  currentUserUsersFollowing: Follower[] | null;
  followingInfo: User[] | null;
  followPhotoFileArray: PostFile[] | null;
  getUsersFollowingConfirm: string | null;
  isLoadingPostData: boolean;
  postMetaDataForUser: PostMetaData | null;
}

const mapStateToProps = createStructuredSelector<AppState, LinkStateProps>({
  currentUser: selectCurrentUser,
  postDataFeedArray: selectPostDataFeedArray,
  postFiles: selectPostFiles,
  postConfirm: selectPostConfirm,
  postError: selectPostError,
  getPostDataConfirm: selectGetPostDataConfirm,
  getPostDataError: selectGetPostDataError,
  getPostFileConfirm: selectGetPostFileConfirm,
  getPostFileError: selectGetPostFileError,
  currentUserUsersFollowing: selectCurrentUserUsersFollowing,
  followingInfo: selectFollowingInfo,
  followPhotoFileArray: selectFollowPhotoFileArray,
  getUsersFollowingConfirm: selectGetUsersFollowingConfirm,
  isLoadingPostData: selectIsLoadingPostData,
  postMetaDataForUser: selectPostMetaDataForUser,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  getPostDataStart: (postDataReq: PostDataReq) =>
    dispatch(getPostDataStart(postDataReq)),
  getPostFileStart: (fileReq: PostFileReq) =>
    dispatch(getPostFileStart(fileReq)),
  clearPostState: () => dispatch(clearPostState()),
  getUsersFollowingStart: (usersFollowingObj: UsersFollowingRequest) =>
    dispatch(getUsersFollowingStart(usersFollowingObj)),
  getOtherUserStart: (otherUserRequest: OtherUserRequest) =>
    dispatch(getOtherUserStart(otherUserRequest)),
  clearFollowersAndFollowing: () => dispatch(clearFollowersAndFollowing()),
  clearFollowState: () => dispatch(clearFollowState()),
});

export default connect(mapStateToProps, mapDispatchToProps)(FeedPage);
