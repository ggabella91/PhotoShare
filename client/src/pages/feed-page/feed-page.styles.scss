@mixin respond-medium {
  @media (max-width: 724px) {
    @content;
  }
}

@mixin respond-wide {
  @media (max-width: 1024px) {
    @content;
  }
}

.feed-page {
  min-height: 800px;
  display: flex;
  width: 85vw;
  flex-direction: column;
  align-items: center;
  margin-left: 50%;
  transform: translate(-50%);

  @include respond-wide() {
    width: 70vw;
    padding: 1vh 0;
  }

  @include respond-medium() {
    width: 95vw;
  }
}

.no-franz {
  margin-top: 15vh;
  font-size: 24px;
  font-weight: 420;
}
