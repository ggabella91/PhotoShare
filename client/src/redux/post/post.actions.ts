import { UserInfoAndOtherData } from '../../components/user-info/user-info.component';
import { PostModalDataToFeed } from '../../components/feed-post-container/feed-post-container.component';

import {
  Post,
  PostDataReq,
  PostFileReq,
  Reaction,
  ReactionReq,
  ArchivePostReq,
  DeleteReactionReq,
  PostFile,
  PostError,
  PostActions,
  PostActionTypes,
  PostMetaData,
} from './post.types';

export const createPostStart = (post: FormData): PostActionTypes => ({
  type: PostActions.CREATE_POST_START,
  payload: post,
});

export const createPostSuccess = (post: Post): PostActionTypes => ({
  type: PostActions.CREATE_POST_SUCCESS,
  payload: post,
});

export const createPostFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.CREATE_POST_FAILURE,
  payload: error,
});

export const createPostReactionStart = (
  reactionReq: ReactionReq
): PostActionTypes => ({
  type: PostActions.CREATE_POST_REACTION_START,
  payload: reactionReq,
});

export const createPostReactionSuccess = (
  reaction: Reaction
): PostActionTypes => ({
  type: PostActions.CREATE_POST_REACTION_SUCCESS,
  payload: reaction,
});

export const createPostReactionFailure = (
  error: PostError
): PostActionTypes => ({
  type: PostActions.CREATE_POST_REACTION_FAILURE,
  payload: error,
});

export const getPostReactionsStart = (postId: string): PostActionTypes => ({
  type: PostActions.GET_POST_REACTIONS_START,
  payload: postId,
});

export const getPostReactionsSuccess = (
  reactions: Reaction[]
): PostActionTypes => ({
  type: PostActions.GET_POST_REACTIONS_SUCCESS,
  payload: reactions,
});

export const getPostReactionsFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.GET_POST_REACTIONS_FAILURE,
  payload: error,
});

export const clearPostStatuses = (): PostActionTypes => ({
  type: PostActions.CLEAR_POST_STATUSES,
  payload: null,
});

export const updateProfilePhotoStart = (photo: FormData): PostActionTypes => ({
  type: PostActions.UPDATE_PROFILE_PHOTO_START,
  payload: photo,
});

export const updateProfilePhotoSuccess = (photo: Post): PostActionTypes => ({
  type: PostActions.UPDATE_PROFILE_PHOTO_SUCCESS,
  payload: photo,
});

export const updateProfilePhotoFailure = (
  error: PostError
): PostActionTypes => ({
  type: PostActions.UPDATE_PROFILE_PHOTO_FAILURE,
  payload: error,
});

export const clearProfilePhotoStatuses = (): PostActionTypes => ({
  type: PostActions.CLEAR_PROFILE_PHOTO_STATUSES,
  payload: null,
});

export const getPostDataStart = (
  postDataReq: PostDataReq
): PostActionTypes => ({
  type: PostActions.GET_POST_DATA_START,
  payload: postDataReq,
});

export const getPostDataSuccess = (postDataArray: Post[]): PostActionTypes => ({
  type: PostActions.GET_POST_DATA_SUCCESS,
  payload: postDataArray,
});

export const addPostDataToFeedArray = (
  postDataArray: Post[]
): PostActionTypes => ({
  type: PostActions.ADD_POST_DATA_TO_FEED_ARRAY,
  payload: postDataArray,
});

export const getPostDataFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.GET_POST_DATA_FAILURE,
  payload: error,
});

export const getPostFileStart = (fileReq: PostFileReq): PostActionTypes => ({
  type: PostActions.GET_POST_FILE_START,
  payload: fileReq,
});

export const getPostFileSuccess = (file: PostFile): PostActionTypes => ({
  type: PostActions.GET_POST_FILE_SUCCESS,
  payload: file,
});

export const getPostFileFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.GET_POST_FILE_FAILURE,
  payload: error,
});

export const getProfilePhotoFileSuccess = (
  file: PostFile
): PostActionTypes => ({
  type: PostActions.GET_PROFILE_PHOTO_FILE_SUCCESS,
  payload: file,
});

export const archivePostStart = (
  archiveReq: ArchivePostReq
): PostActionTypes => ({
  type: PostActions.ARCHIVE_POST_START,
  payload: archiveReq,
});

export const archivePostSuccess = (message: string): PostActionTypes => ({
  type: PostActions.ARCHIVE_POST_SUCCESS,
  payload: message,
});

export const archivePostFailure = (error: Error): PostActionTypes => ({
  type: PostActions.ARCHIVE_POST_FAILURE,
  payload: error,
});

export const deleteReactionStart = (
  deleteReactionReq: DeleteReactionReq
): PostActionTypes => ({
  type: PostActions.DELETE_REACTION_START,
  payload: deleteReactionReq,
});

export const deleteReactionSuccess = (message: string): PostActionTypes => ({
  type: PostActions.DELETE_REACTION_SUCCESS,
  payload: message,
});

export const deleteReactionFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.DELETE_REACTION_FAILURE,
  payload: error,
});

export const getOtherUserProfilePhotoFileSuccess = (
  file: string
): PostActionTypes => ({
  type: PostActions.GET_OTHER_USER_PROFILE_PHOTO_FILE_SUCCESS,
  payload: file,
});

export const getUserPhotoForFollowArraySuccess = (
  postFile: PostFile
): PostActionTypes => ({
  type: PostActions.GET_USER_PHOTO_FOR_FOLLOW_ARRAY_SUCCESS,
  payload: postFile,
});

export const getUserPhotoForSuggestionArraySuccess = (
  postFile: PostFile
): PostActionTypes => ({
  type: PostActions.GET_USER_PHOTO_FOR_SUGGESTION_ARRAY_SUCCESS,
  payload: postFile,
});

export const getUserPhotoForReactorArraySuccess = (
  postFile: PostFile
): PostActionTypes => ({
  type: PostActions.GET_USER_PHOTO_FOR_REACTOR_ARRAY_SUCCESS,
  payload: postFile,
});

export const clearArchivePostStatuses = (): PostActionTypes => ({
  type: PostActions.CLEAR_ARCHIVE_POST_STATUSES,
  payload: null,
});

export const clearFollowPhotoFileArray = (): PostActionTypes => ({
  type: PostActions.CLEAR_FOLLOW_PHOTO_FILE_ARRAY,
  payload: null,
});

export const clearSuggestionPhotoFileArray = (): PostActionTypes => ({
  type: PostActions.CLEAR_SUGGESTION_PHOTO_FILE_ARRAY,
  payload: null,
});

export const clearPostFilesAndData = (): PostActionTypes => ({
  type: PostActions.CLEAR_POST_FILES_AND_DATA,
  payload: null,
});

export const clearPostState = (): PostActionTypes => ({
  type: PostActions.CLEAR_POST_STATE,
  payload: null,
});

export const setCommentToDelete = (
  deleteReactionReq: DeleteReactionReq
): PostActionTypes => ({
  type: PostActions.SET_COMMENT_TO_DELETE,
  payload: deleteReactionReq,
});

export const setShowCommentOptionsModal = (
  showCommentOptionsModal: boolean
): PostActionTypes => ({
  type: PostActions.SET_SHOW_COMMENT_OPTIONS_MODAL,
  payload: showCommentOptionsModal,
});

export const clearPostReactions = (): PostActionTypes => ({
  type: PostActions.CLEAR_POST_REACTIONS,
  payload: null,
});

export const setPostMetaDataForUser = (
  postMetaData: PostMetaData
): PostActionTypes => ({
  type: PostActions.SET_POST_META_DATA_FOR_USER,
  payload: postMetaData,
});

export const setPostLikingUsersArray = (
  postLikingUsersArray: UserInfoAndOtherData[]
): PostActionTypes => ({
  type: PostActions.SET_POST_LIKING_USERS_ARRAY,
  payload: postLikingUsersArray,
});

export const setShowPostLikingUsersModal = (
  showPostLikingUsersModal: boolean
): PostActionTypes => ({
  type: PostActions.SET_SHOW_POST_LIKING_USERS_MODAL,
  payload: showPostLikingUsersModal,
});

export const setFeedPagePostModalData = (
  postModalDataToFeed: PostModalDataToFeed
): PostActionTypes => ({
  type: PostActions.SET_FEED_PAGE_POST_MODAL_DATA,
  payload: postModalDataToFeed,
});

export const setFeedPagePostModalShow = (
  postModalDataShow: boolean
): PostActionTypes => ({
  type: PostActions.SET_FEED_PAGE_POST_MODAL_SHOW,
  payload: postModalDataShow,
});

export const setFeedPagePostOptionsModalShow = (
  postOptionsModalDataShow: boolean
): PostActionTypes => ({
  type: PostActions.SET_FEED_PAGE_POST_OPTIONS_MODAL_SHOW,
  payload: postOptionsModalDataShow,
});

export const setClearFeedPagePostModalState = (
  clearFeedPagePostModalState: boolean
): PostActionTypes => ({
  type: PostActions.SET_CLEAR_FEED_PAGE_POST_MODAL_STATE,
  payload: clearFeedPagePostModalState,
});

// Actions for feed-post-container

export const getFeedPostFileSuccess = (file: PostFile): PostActionTypes => ({
  type: PostActions.GET_FEED_POST_FILE_SUCCESS,
  payload: file,
});

export const getFeedPostReactionsSuccess = (
  reactions: Reaction[]
): PostActionTypes => ({
  type: PostActions.GET_FEED_POST_REACTIONS_SUCCESS,
  payload: reactions,
});

export const getUserPhotoForFeedReactorArraySuccess = (
  postFile: PostFile
): PostActionTypes => ({
  type: PostActions.GET_USER_PHOTO_FOR_FEED_REACTOR_ARRAY_SUCCESS,
  payload: postFile,
});
