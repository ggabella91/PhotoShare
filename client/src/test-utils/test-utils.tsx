import { FC, ReactElement } from 'react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import { render, RenderOptions } from '@testing-library/react';
import { store } from '../redux/store';

interface WrapperProps {
  wrapperProps?: { route: string; location: string };
}

type CustomRenderOptions = RenderOptions & WrapperProps;

const WithReduxProvider: FC<WrapperProps> = ({ children, wrapperProps }) => {
  return (
    <Provider store={store}>
      <BrowserRouter>
        <Routes>
          <Route path={wrapperProps?.route || '/'} element={children} />
        </Routes>
      </BrowserRouter>
    </Provider>
  );
};

const renderWithReduxAndRouter = (
  ui: ReactElement,
  options?: Omit<CustomRenderOptions, 'wrapper'>
) => {
  window.history.pushState({}, '', options?.wrapperProps?.location || '/');
  const customContainer = document.createElement('div');
  customContainer.id = 'root';

  return render(ui, {
    wrapper: (props) => (
      <WithReduxProvider {...props} wrapperProps={options?.wrapperProps} />
    ),
    container: document.body.appendChild(customContainer),
    ...options,
  });
};

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
export { renderWithReduxAndRouter as render };
