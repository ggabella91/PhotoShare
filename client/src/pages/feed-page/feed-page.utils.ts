import { Follower } from '../../redux/follower/follower.types';
import { User } from '../../redux/user/user.types';
import { Post, PostFile, Reaction } from '../../redux/post/post.types';
import { List, Map } from 'immutable';

import { UserInfoAndPostFile } from './feed-page.component';
import { UserInfoAndOtherData } from '../../components/user-info/user-info.component';

export const prepareUserInfoAndFileArray = (
  followingInfoList: List<User>,
  dataFeedList: List<List<Post>>,
  followingProfilePhotoList: List<PostFile>,
  postFileFeedList: List<PostFile>
) => {
  let userInfoAndPostObjArray: List<UserInfoAndPostFile> = postFileFeedList.map(
    (el) => {
      let location: string;
      let dateString: string;
      let dateInt: number;
      let id: string;
      let username: string;
      let postId: string;
      let postS3Key: string;
      let profilePhotoS3Key: string;
      let profilePhotoString: string;
      let caption: string;

      dataFeedList.forEach((innerList) => {
        innerList.forEach((innerEl) => {
          if (innerEl.s3Key === el.s3Key) {
            let date = innerEl.createdAt;

            location = innerEl.postLocation || '';
            id = innerEl.userId;
            postId = innerEl.id;
            postS3Key = innerEl.s3Key;
            caption = innerEl.caption || '';
            dateString = new Date(date).toDateString();
            dateInt = new Date(date).getTime();
          }
        });
      });

      followingInfoList.forEach((userEl) => {
        if (userEl.id === id!) {
          username = userEl.username;
          profilePhotoS3Key = userEl.photo || '';
        }
      });

      followingProfilePhotoList.forEach((userEl) => {
        if (profilePhotoS3Key! && userEl.s3Key === profilePhotoS3Key) {
          profilePhotoString = userEl.fileString;
        }
      });

      if (!profilePhotoString!) {
        profilePhotoString = '';
      }
      if (!location!) {
        location = '';
      }

      if (!caption!) {
        caption = '';
      }

      if (!dateString!) {
        dateString = '';
      }

      if (!dateInt!) {
        dateInt = Date.now();
      }

      return {
        username: username!,
        userId: id!,
        profilePhotoFileString: profilePhotoString,
        location,
        postId: postId!,
        postS3Key: postS3Key!,
        postFileString: el.fileString,
        caption,
        dateString,
        dateInt,
      };
    }
  );

  return userInfoAndPostObjArray;
};

export const compareFollowerArrays = (
  array1: Follower[],
  array2: Follower[]
) => {
  if (array1.length !== array2.length) {
    return false;
  }

  for (let i = 0; i < array1.length; i++) {
    if (array1[i].followerId !== array2[i].followerId) {
      return false;
    }
  }

  return true;
};

export const compareUserOrPostOrReactionLists = (
  list1: List<User> | List<Post> | List<Reaction>,
  list2: List<User> | List<Post> | List<Reaction>
) => {
  if (list1.size !== list2.size) {
    return false;
  }

  for (let i = 0; i < list1.size; i++) {
    if (list1.get(i)!.id !== list2.get(i)!.id) {
      return false;
    }
  }

  return true;
};

export const comparePostFileLists = (
  list1: List<PostFile>,
  list2: List<PostFile>
) => {
  if (list1.size !== list2.size) {
    return false;
  }

  for (let i = 0; i < list1.size; i++) {
    if (list1.get(i)!.s3Key !== list2.get(i)!.s3Key) {
      return false;
    }
  }

  return true;
};

export const compareUserInfoAndDataObjArrays = (
  array1: UserInfoAndPostFile[] | UserInfoAndOtherData[],
  array2: UserInfoAndPostFile[] | UserInfoAndOtherData[]
) => {
  if (array1.length !== array2.length) {
    return false;
  }

  for (let i = 0; i < array1.length; i++) {
    let array1AtIdxValues = Object.values(array1[i]);
    let array2AtIdxValues = Object.values(array2[i]);

    for (let j = 0; j < array1AtIdxValues.length; j++) {
      if (array1AtIdxValues[j] !== array2AtIdxValues[j]) {
        return false;
      }
    }
  }

  return true;
};
