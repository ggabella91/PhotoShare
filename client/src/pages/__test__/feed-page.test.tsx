import { render, screen } from '../../test-utils/test-utils';

import { FeedPage } from '../feed-page/feed-page.component';

import { DeleteReactionReq } from '../../redux/post/post.types';

describe('feed page component tests', () => {
  const setup = () => {
    const commentToDelete = {} as DeleteReactionReq;
    const clearFollowState = jest.fn();
    const clearFollowersAndFollowing = jest.fn();
    const clearPostState = jest.fn();
    const getPostDataStart = jest.fn();
    const getPostFileStart = jest.fn();
    const getOtherUserStart = jest.fn();
    const getUsersFollowingStart = jest.fn();
    const archivePostStart = jest.fn();
    const setShowPostLikingUsersModal = jest.fn();
    const setFeedPagePostModalShow = jest.fn();
    const setFeedPagePostOptionsModalShow = jest.fn();
    const setClearFeedPagePostModalState = jest.fn();
    const setShowCommentOptionsModal = jest.fn();
    const deleteReactionStart = jest.fn();
    const mockIntersectionObserver = jest.fn();
    mockIntersectionObserver.mockReturnValue({
      observe: jest.fn(),
      disconnect: jest.fn(),
    });
    window.IntersectionObserver = mockIntersectionObserver;

    render(
      <FeedPage
        clearFollowState={clearFollowState}
        clearFollowersAndFollowing={clearFollowersAndFollowing}
        clearPostState={clearPostState}
        getPostDataStart={getPostDataStart}
        getPostFileStart={getPostFileStart}
        getOtherUserStart={getOtherUserStart}
        getUsersFollowingStart={getUsersFollowingStart}
        feedPagePostModalShow={false}
        feedPagePostOptionsModalShow={false}
        clearFeedPagePostModalState={false}
        showCommentOptionsModal={false}
        commentToDelete={commentToDelete}
        archivePostStart={archivePostStart}
        setShowPostLikingUsersModal={setShowPostLikingUsersModal}
        setFeedPagePostModalShow={setFeedPagePostModalShow}
        setFeedPagePostOptionsModalShow={setFeedPagePostOptionsModalShow}
        setClearFeedPagePostModalState={setClearFeedPagePostModalState}
        setShowCommentOptionsModal={setShowCommentOptionsModal}
        deleteReactionStart={deleteReactionStart}
      />
    );

    return {
      clearFollowState,
      clearFollowersAndFollowing,
      clearPostState,
      getPostDataStart,
      getPostFileStart,
      getOtherUserStart,
      getUsersFollowingStart,
      archivePostStart,
      setShowPostLikingUsersModal,
      setFeedPagePostModalShow,
      setFeedPagePostOptionsModalShow,
      setClearFeedPagePostModalState,
      setShowCommentOptionsModal,
      deleteReactionStart,
    };
  };

  it('render a feed page component', () => {
    const {
      getPostDataStart,
      getPostFileStart,
      getOtherUserStart,
      getUsersFollowingStart,
    } = setup();

    const feedPage = screen.getByTestId('feed-page');

    const feedPostContainers = screen.getAllByTestId('feed-post-container');

    expect(feedPage).toBeInTheDocument();
    expect(getPostDataStart).toBeCalled();
    expect(getPostFileStart).toBeCalled();
    expect(getOtherUserStart).toBeCalled();
    expect(getUsersFollowingStart).toBeCalled();
    expect(feedPostContainers.length).toEqual(4);
  });
});
