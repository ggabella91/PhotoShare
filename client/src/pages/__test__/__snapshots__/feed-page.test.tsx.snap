// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a feed page component 1`] = `
<div
  className="feed-page"
>
  <div
    className="no-franz"
  >
    Follow users to see their recent posts here
  </div>
</div>
`;
